### PR TITLE
Support for creating and updating ICS 08 WASM clients wrapping an ICS 07 Tendermint client

### PR DESCRIPTION
**Pull request by romac**
Part of: #3951

Depends on: #3942

### 1. Get `simd` with Wasm client support

From the root of the `hermes` repository, on this branch, run:

```bash
❯ nix shell '.#ibc-go-v7-wasm-simapp'
❯ which simd
/nix/store/3pp59wi249yd30dmr7hb49wbm9s19zci-simd-v7.3.0-wasm/bin/simd
```

### 2. Setup two `simd` chains with Wasm support

Put this into `./gm.toml`, update `gaiad_binary` to point to the `simd` binary we installed above, and update the path to the `hermes` binary:

```toml
[global]
home_dir             = "$HOME/.gm"
gaiad_binary         = "/nix/store/3pp59wi249yd30dmr7hb49wbm9s19zci-simd-v7.3.0-wasm/bin/simd"
extra_wallets        = 1
auto_maintain_config = true
add_to_hermes        = true

[global.hermes]
binary            = "$HOME/Informal/Code/hermes/target/debug/hermes"
config            = "$HOME/.hermes/config.toml"
log_level         = "debug"
telemetry_enabled = false
telemetry_host    = "127.0.0.1"
telemetry_port    = 3001

[ibc0-wasm]
ports_start_at = 27000

[ibc1-wasm]
ports_start_at = 28000
```

### 3. Create the chains and update their config

Create, then stop and re-initialize the chains:

```bash
❯ GM_TOML=./gm.toml gm start
❯ GM_TOML=./gm.toml gm stop
❯ GM_TOML=./gm.toml gm nuke
```

Update the config of the chain `ibc0-wasm`:

- Fetch the scripts from this Gist: https://gist.github.com/romac/80eddfd304ec03562badf23d9b43f892

- Run `enable-was.py`:
  ```bash
  $ p3 enable-wasm.py ~/.gm/ibc0-wasm/
  ```
- Or do it manually:
  - In `~/.gm/ibc0-wasm/config/config.toml`, change `max_body_bytes` to `10001048576`
  - In `~/.gm/ibc0-wasm/config/genesis.json`, change:
    - `max_deposit_period` to `10s`
    - `voting_period` to `10s`
    - Add `08-wasm` to the list of alloed clients in `allowed_clients`

```diff
@@ -246,8 +246,8 @@
             "amount": "10000000"
           }
         ],
-        "max_deposit_period": "172800s",
-        "voting_period": "172800s",
+        "max_deposit_period": "10s",
+        "voting_period": "10s",
         "quorum": "0.334000000000000000",
         "threshold": "0.500000000000000000",
         "veto_threshold": "0.334000000000000000",
@@ -276,6 +276,7 @@
           "allowed_clients": [
             "06-solomachine",
             "07-tendermint",
+            "08-wasm",
             "09-localhost"
           ]
         },
```

### 4. Start the chains

```bash
❯ GM_TOML=./gm.toml hermes config
❯ GM_TOML=./gm.toml hermes keys
❯ GM_TOML=./gm.toml gm start
```

### 5. Install the contract on chain

- Get a wallet address of the `validator` account for `ibc0-wasm`

```bash
❯ GM_TOML=./gm.toml gm keys ibc0-wasm
- address: cosmos15culynpgfdcfupl3f2yrzmav4t8hhpfalc2eu6
  name: validator
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"A4qOppP8gT3fF/671w7qKBFllT0DTlzwawwZ9b2dx926"}'
```

- Run the install script from the Gist:

```bash
❯ python3 ./install-wasm.py ibc0-wasm tcp://localhost:27000 cosmos15culynpgfdcfupl3f2yrzmav4t8hhpfalc2eu6 tools/integration-test/fixtures/wasm/ibc_client_tendermint_cw.wasm
```

### 6. Create a Wasm client using Hermes

```
❯ cargo run -- create client --host-chain ibc0-wasm --reference-chain ibc1-wasm --wasm-checksum 08eb0025bc235c3114946ab5124ade365061c4529068dcae180357e819a7cb07
SUCCESS CreateClient(
    CreateClient(
        Attributes {
            client_id: ClientId(
                "08-wasm-0",
            ),
            client_type: Wasm,
            consensus_height: Height {
                revision: 0,
                height: 220,
            },
        },
    ),
)
```

### 7. Update the Wasm client using Hermes

```
❯ cargo run -- update client --host-chain ibc0-wasm --client 08-wasm-0
SUCCESS [
    UpdateClient(
        UpdateClient {
            common: Attributes {
                client_id: ClientId(
                    "08-wasm-0",
                ),
                client_type: Wasm,
                consensus_height: Height {
                    revision: 0,
                    height: 232,
                },
            },
            header: None,
        },
    ),
]
```

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by romac** [comment]
CI consistently fails with

> Client refresh failed: error raised while updating client on chain ibc1: failed sending message to dst chain
> 
> Caused by:
>     gRPC call `send_tx_simulate` failed with status: status: Unknown, message: "recovered: IBC validation/execution context error: ICS02 Client error: other error: `cannot convert into a `Height` type from string `\0\0\0\0\0\0\0-``: wasm contract call failed [cosmos/ibc-go/modules/light-clients/08-wasm/types/vm.go:156]

whereas the test goes through successfully locally on macOS.

I was able to reproduce the failure, though not consistently, by tweaking the `trusting_period` of the clients. I wonder if there is a race condition where the client expires after we have sent the `UpdateClient` message, which ends up triggering the error above within the Wasm contract?

**Comment by romac** [comment]
Alright, increasing the trusting period on CI fixed it, no idea why though…

File: tools/integration-test/src/tests/wasm_client.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        info!("Creating client A to B");
+        let mut client_a_to_b =
+            bootstrap_foreign_client(&chain_a, &chain_b, overrides.client_options_a_to_b())?;
+
+        info!("Creating client B to A");
+        let mut client_b_to_a =
+            bootstrap_foreign_client(&chain_b, &chain_a, overrides.client_options_b_to_a())?;
+
+        info!("Refreshing client A to B...");
+        let res = client_a_to_b.refresh(true);
```
Comment (ljoss17) [inline tools/integration-test/src/tests/wasm_client.rs:117]: Could we use `client_refresh_rate` and wait for the appropriate amount of time https://github.com/informalsystems/hermes/blob/v1.8.2/config.toml#L356?
Related to the comment under `try_refresh()`

**Review by ljoss17 — COMMENTED** [review]
Looks great, thanks for this!!
I left a few nits and questions

File: crates/relayer/src/foreign_client.rs [hunk 2]
```diff
@@ -912,18 +951,15 @@ impl<DstChain: ChainHandle, SrcChain: ChainHandle> ForeignClient<DstChain, SrcCh
...
-                    info!(?elapsed, ?refresh_period, "client needs to be refreshed");
+        let elapsed = elapsed.unwrap_or_default();
 
-                    self.build_latest_update_client_and_send()
-                        .map_or_else(Err, |ev| Ok(Some(ev)))
-                } else {
-                    Ok(None)
-                }
-            }
+        if force || elapsed > refresh_period {
```
[the remark this answers is not in the document]
Comment (romac) [reply crates/relayer/src/foreign_client.rs:956]: That's what I was doing initially but the purpose of this test is not to test whether or not the client is refreshed or not when needed but just to update so I thought I'd add a way to force the refresh. But yeah it affects call sites which don't care about that so I think what I'll do instead is add a `force_refresh` method.

File: tools/integration-test/src/tests/wasm_client.rs (hunk 1 above)
Comment (romac) [reply tools/integration-test/src/tests/wasm_client.rs:109]: See my comment above.

File: crates/relayer-types/src/clients/ics08_wasm/error.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        MissingLatestHeight
+            |_| { "missing latest height" },
+
+        InvalidRawHeight
+            { raw_height: RawHeight }
+            |e| { format_args!("invalid raw height: {:?}", e.raw_height) },
+
+        ProtoEncode
+            [ TraceError<prost::EncodeError> ]
+            | _ | { "Protobuf encoding error error" },
```
[the remark this answers is not in the document]
Comment (romac) [reply crates/relayer-types/src/clients/ics08_wasm/error.rs]: https://github.com/informalsystems/hermes/pull/3943/commits/b06cb13c8a38aba7f8d92dded238f1cd1e293455

File: crates/relayer-types/src/clients/ics08_wasm/client_message.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+impl Protobuf<RawClientMessage> for ClientMessage {}
+
+impl TryFrom<RawClientMessage> for ClientMessage {
+    type Error = Error;
+
+    fn try_from(value: RawClientMessage) -> Result<Self, Self::Error> {
+        let any = Any::decode(value.data.as_slice()).map_err(Error::proto_decode)?;
+        let data = AnyHeader::try_from(any).unwrap();
```
[the remark this answers is not in the document]
Comment (romac) [reply crates/relayer-types/src/clients/ics08_wasm/client_message.rs]: https://github.com/informalsystems/hermes/pull/3943/commits/b06cb13c8a38aba7f8d92dded238f1cd1e293455

**Review by ljoss17 — APPROVED** [review]
Looks good to me!

**Pull request merged**